crates/runtime/src/embeddings/cosine_distance.rs

```diff
@@ -0,0 +1,259 @@
+
+macro_rules! downcast_arg {
+    ($ARG:expr, $ARRAY_TYPE:ident) => {{
+        $ARG.as_any().downcast_ref::<$ARRAY_TYPE>().ok_or_else(|| {
+            DataFusionError::Internal(format!("could not cast to {}", type_name::<$ARRAY_TYPE>()))
+        })?
+    }};
+}
+
+/// array function wrapper that differentiates between scalar (length 1) and array.
+pub(crate) fn make_scalar_function<F>(
+    inner: F,
+) -> impl Fn(&[ColumnarValue]) -> DataFusionResult<ColumnarValue>
+where
+    F: Fn(&[ArrayRef]) -> DataFusionResult<ArrayRef>,
+{
+    move |args: &[ColumnarValue]| {
+        // first, identify if any of the arguments is an Array. If yes, store its `len`,
+        // as any scalar will need to be converted to an array of len `len`.
+        let len = args
+            .iter()
+            .fold(Option::<usize>::None, |acc, arg| match arg {
+                ColumnarValue::Scalar(_) => acc,
+                ColumnarValue::Array(a) => Some(a.len()),
+            });
+
+        let args = ColumnarValue::values_to_arrays(args)?;
+
+        let result = (inner)(&args);
+
+        // If all inputs are scalar, keeps output as scalar
+        if len.is_none() {
+            let result = result.and_then(|arr| ScalarValue::try_from_array(&arr, 0));
+            result.map(ColumnarValue::Scalar)
+        } else {
+            result.map(ColumnarValue::Array)
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct CosineDistance {
+    signature: Signature,
+}
+impl Default for CosineDistance {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl CosineDistance {
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            signature: Signature::user_defined(Volatility::Immutable),
+        }
+    }
+}
+
+impl ScalarUDFImpl for CosineDistance {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn name(&self) -> &str {
+        "cosine_distance"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, arg_types: &[DataType]) -> DataFusionResult<DataType> {
+        match arg_types[0] {
+            List(_) | LargeList(_) | FixedSizeList(_, _) => Ok(Float64),
+            _ => exec_err!(
+                "The cosine_distance function can only accept List/LargeList/FixedSizeList."
+            ),
+        }
+    }
+
+    fn coerce_types(&self, arg_types: &[DataType]) -> DataFusionResult<Vec<DataType>> {
+        if arg_types.len() != 2 {
+            return exec_err!("cosine_distance expects exactly two arguments");
+        }
+        let mut result = Vec::new();
+        for arg_type in arg_types {
+            match arg_type {
+                List(_) | LargeList(_) | FixedSizeList(_, _) => {
+                    result.push(coerced_fixed_size_list_to_list(arg_type));
+                }
+                _ => {
+                    return exec_err!(
+                    "The cosine_distance function can only accept List/LargeList/FixedSizeList."
+                )
```

Comment:
```suggestion
                        "The cosine_distance function can only accept List/LargeList/FixedSizeList."
                    )
```

Reply:
Odd `cargo fmt` would be okay with this.